chainer/function_node.py

```diff
@@ -747,55 +746,57 @@ def grad(outputs, inputs, grad_outputs=None, grad_inputs=None, set_grad=False,
             'grad_inputs must be a tuple or a list or None, not {}.'.format(
                 type(grad_inputs)))
 
-    for v in outputs:
-        # Raise error here if v is created by Function.backward.
-        # In such case, we don't know exact inputs of the creator.
-        v.node._check_old_style_gradient()
-
     # The implementation consists of three steps.
 
     # 1. Backward enumeration: all the nodes reachable backward from the output
     #    nodes are enumerated. The forward direction links are collected in
     #    this step. Note that the variable nodes whose requires_grad is false
     #    are ignored and their creators are not searched.
-    candidate_funcs = [v.creator_node for v in outputs
-                       if v.creator_node is not None]
-    visited_funcs = set()
-    forward_graph = collections.defaultdict(list)
-    while candidate_funcs:
-        func = candidate_funcs.pop()
-        if func in visited_funcs:
+    candidate_vars = [y.node for y in outputs]
+    output_nodes = list(candidate_vars)  # TODO(kataoka): remove this
+    visited_vars = set()
+    forward_var_to_func = collections.defaultdict(list)
+    forward_func_to_var = collections.defaultdict(list)
+    while candidate_vars:
+        y = candidate_vars.pop()
+        if y in visited_vars:
             continue
-        visited_funcs.add(func)
-        for x in func.inputs:
-            # Raise error here if x is created by Function.backward.
-            # In such case, we don't know exact inputs of the creator.
-            x._check_old_style_gradient()
+        visited_vars.add(y)
 
-            if not x.requires_grad:
-                continue
-            forward_graph[x].append(func)
-            creator = x.creator_node
-            if creator is not None and creator not in visited_funcs:
-                candidate_funcs.append(creator)
+        # Raise error here if y is created by Function.backward.
+        # In such case, we don't know exact inputs of the creator.
+        y._check_old_style_gradient()
+
+        func = y.creator_node
+        if func is None:
+            continue
+        if func not in forward_func_to_var:  # func is not visited
+            for x in func.inputs:
+                if x.requires_grad:
+                    candidate_vars.append(x)
```

Comment:
How about checking `x not in visited_vars` before `candidate_vars.append(x)`?
(Instead of checking `y in visited_vars` after dequeuing?)